### PR TITLE
sys-kernel/installkernel-systemd-boot: revbump with machine-id dir fix

### PR DESCRIPTION
**Pull request by gyakovlev**
helpful links

https://github.com/systemd/systemd/blob/7ad9bad71b077f6a4629f4e48fb9ad26e1784809/src/kernel-install/kernel-install#L88-L95

https://github.com/systemd/systemd/blob/7ad9bad71b077f6a4629f4e48fb9ad26e1784809/src/kernel-install/kernel-install#L102-L114

and whole https://github.com/systemd/systemd/blob/main/src/kernel-install/00-entry-directory.install

### CODE REVIEW

**Comment by gyakovlev** [comment]
as alternative, we can install it as `/usr/lib/kernel/install.d/` file, something before to be processed before`00-entry-directory.install`.

in that case I think it can re-use `kernel_install's vars directly, without doing all the search again.

why do best ideas come later?

File: sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+		kernel_dir="/boot/${machine_id}"
+	elif [[ -d /boot/efi/loader/entries ]] || [[ -d /boot/efi/${machine_id} ]]; then
+		kernel_dir="/boot/efi/${machine_id}"
+	elif mountpoint -q /efi; then
+		kernel_dir="/efi/${machine_id}"
+	elif mountpoint -q /boot/efi; then
+		kernel_dir="/boot/efi/${machine_id}"
+	else
+		kernel_dir="/boot/${machine_id}"
+	fi
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh:29]: I suppose the different /boot/efi vs /efi order comes from kernel-install itself?

File: sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+create_kernel_dir() {
+	# logic flow taken from kernel-install itself
+	local machine_id kernel_dir
+
+	if [[ -d /efi/Default ]] || [[ -d /boot/Default ]] || [[ -d /boot/efi/Default ]]; then
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh:7]: If you're going for bash, I suppose you can put the `||`s inside `[[ ... ]]`.

File: sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		machine_id="Default"
+	elif [[ -f /etc/machine-id ]]; then
+		read machine_id < /etc/machine-id
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh:10]: The script's going to go nuts if this fails… or the file is empty.

File: sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	elif mountpoint -q /efi; then
+		kernel_dir="/efi/${machine_id}"
+	elif mountpoint -q /boot/efi; then
+		kernel_dir="/boot/efi/${machine_id}"
+	else
+		kernel_dir="/boot/${machine_id}"
+	fi
+
+	if [[ ! -d ${kernel_dir} ]]; then
+		mkdir -p "${kernel_dir}" || { echo "Could not create ${kernel_dir} directory" 1>&2 ; exit 1; }
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh:32]: Please use `if` here.

File: sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	fi
+}
+
+	create_kernel_dir
+	exec kernel-install add "\${1}" "\${2}"
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/installkernel-0.0.1.sh:37]: The indentation looks wrong.

**Review by mgorny — CHANGES_REQUESTED** [review]

**Comment by Zlogene** [comment]
@gyakovlev please push straight to stable as we are now having totally b0rken version

File: sys-kernel/installkernel-systemd-boot/files/00-00machineid-directory.install [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+if ! [[ $KERNEL_INSTALL_MACHINE_ID ]]; then
+    exit 0
+fi
+
+if [[ $COMMAND != add ]]; then
+    exit 0
+fi
+
+# If the machine-id dir does not exist (e.g. $ESP/<machine-id>)
+# create it. It recieves values directly from kernel-install.
```
Comment (SoapGentoo) [inline sys-kernel/installkernel-systemd-boot/files/00-00machineid-directory.install:15]:
```suggestion
# create it. It receives values directly from kernel-install.
```

File: sys-kernel/installkernel-systemd-boot/files/00-00machineid-directory.install [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+# This is the only function of this plugin.
+MACHINE_ID_DIR="${ENTRY_DIR_ABS%/*}"
+if ! [[ -d "${MACHINE_ID_DIR}" ]]; then
+	if [[ "${KERNEL_INSTALL_VERBOSE}" -gt 0 ]]; then
+    		echo "+mkdir -v -p ${MACHINE_ID_DIR}"
+    		exec mkdir -v -p "${MACHINE_ID_DIR}"
+	else
+    		exec mkdir -p "${MACHINE_ID_DIR}"
```
Comment (SoapGentoo) [inline sys-kernel/installkernel-systemd-boot/files/00-00machineid-directory.install:23]: why the exec?

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

File: sys-kernel/installkernel-systemd-boot/files/00-00machineid-directory.install (hunk 7 above)
Comment (gyakovlev) [reply sys-kernel/installkernel-systemd-boot/files/00-00machineid-directory.install:23]: same as in other plugins, no other reason

File: sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-0-r1.ebuild [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+RDEPEND="|| ( sys-apps/systemd sys-boot/systemd-boot )
+	!<sys-apps/debianutils-4.9-r1[installkernel(+)]
+	!sys-kernel/installkernel-gentoo"
+
+src_install() {
+	# we could technically use a symlink here but it would require
+	# us to know the correct path, and that implies /usr merge problems
+	into /
+	newsbin - installkernel <<-EOF
+		#!/bin/sh
```
Comment (SoapGentoo) [inline sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-0-r1.ebuild:24]:
```suggestion
		#!/usr/bin/env sh
```

File: sys-kernel/installkernel-systemd-boot/files/00-00machineid-directory.install [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
Comment (SoapGentoo) [inline sys-kernel/installkernel-systemd-boot/files/00-00machineid-directory.install:1]:
```suggestion
#!/usr/bin/env bash
```

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

**Review by mgorny — APPROVED** [review]
If it works, ship it!

**Pull request closed without merging**